src/Map.jl

```diff
@@ -0,0 +1,476 @@
+"""
+    AbstractMap{D,CoD}
+
+Abstract type for maps between elements of sets.
+Every map has a domain of type `D`, which is the set to which inputs
+belong and a codomain `CoD`, which is the set to which outputs belong. Note
+that maps are not required to be total. That is, the true (co)domain may be
+a subset of the provided (co)domain.
+
+Every new map type must implement [`domain`](@ref) and [`codomain`](@ref) and be
+callable. Optionally, [`inv`](@ref), [`linv`](@ref), and [`rinv`](@ref) may be
+used to specify a map that is a true, left-, or right-inverse, respectively.
+
+Maps can be composed in the same way as functions using the `∘` command. See
+[`CompositeMap`](@ref).
+"""
+abstract type AbstractMap{D,CoD} end
+
+function show(io::IO, mime::MIME"text/plain", m::AbstractMap)
+    print(io, "$(typeof(m).name): 𝑀 ⟶ 𝑁\n",
+              "𝑀 = $(repr(mime, domain(m)))\n",
+              "𝑁 = $(repr(mime, codomain(m)))")
+end
+
+"""
+    domain(m::AbstractMap)
+
+Get the manifold to which inputs to the map `m` belong. By default, this is
+assumed to be stored in the field `m.domain`.
+"""
+domain(m::AbstractMap) = m.domain
+domain(::AbstractMap{M}) where {M<:Euclidean} = M()
+
+"""
+    codomain(m::AbstractMap)
+
+Get the manifold to which outputs to the map `m` belong. By default, this is
+assumed to be stored in the field `m.codomain`.
+"""
+codomain(m::AbstractMap) = m.codomain
+codomain(::AbstractMap{M,N}) where {M,N<:Euclidean} = N()
+
+@doc doc"""
+    inv(f::AbstractMap)
+
+For a _total_ map $f \colon M \to M$, return its true inverse $g$, such that
+$g \circ f = f \circ g = \mathrm{id}_M \colon M \to M$, where $\mathrm{id}_M$
+is the identity map on $M$.
+"""
+function inv end
+
+@doc doc"""
+    linv(f::AbstractMap)
+
+For a map $f \colon M \to N$, return its left inverse $g$, such that
+$g \circ f = \mathrm{id}_N \colon N \to N$, where $\mathrm{id}_N$ is the
+identity map on $N$.
+"""
+linv(f::AbstractMap) = inv(f)
+
+@doc doc"""
+    rinv(f::AbstractMap)
+
+For a map $f \colon M \to N$, return its right inverse $g$, such that
+$f \circ g = \mathrm{id}_M \colon M \to M$, where $\mathrm{id}_M$ is the
+identity map on $M$.
+"""
+rinv(f::AbstractMap) = inv(f)
+
+@doc doc"""
+    AbstractCurve{M} = AbstractMap{Euclidean{Tuple{1}},M}
+
+An alias for a curve, a map from 1-D real space to a manifold, i.e.
+$\phi \colon \mathbb R \to M$
+"""
+AbstractCurve{M} = AbstractMap{RealScalars,M}
+
+@doc doc"""
+    AbstractField{M,T<:Tuple} = AbstractMap{M,Euclidean{T}}
+
+An alias for a generic field, a map from a point on a manifold `M` to real
+space of some dimension stored in `T`,
+i.e. $\phi \colon M \to \mathbb R^{m \times \dots \times n}$
+"""
+AbstractField{M,T<:Tuple} = AbstractMap{M,Euclidean{T}}
```

Comment:
This is somewhat questionable, not all fields have to be represented by values on `Euclidean`. I'd say that any Hilbert space would do (and maybe even this is too restrictive). `AbstractField` can be useful but I'd leave codomain up to the user.